interfaces/builtin/modem_manager.go

```diff
@@ -0,0 +1,1223 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var modemManagerPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the ModemManager service. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+# To check present devices
+/run/udev/data/* r,
+/sys/bus/usb/devices/ r,
+# FIXME snapd should be querying udev and adding the /sys and /run/udev accesses
+# that are assigned to the snap, but we are not there yet.
```

Comment:
The idea is this:
1. if snapd knows the device in /dev, do `udevadm info /dev/foo` otherwise if snapd knows the device in /sys/devices, do `udevadm info /sys/devices/.../foo`. Eg:
```
$ udevadm info /dev/kmsg
P: /devices/virtual/mem/kmsg
N: kmsg
E: DEVMODE=0644
E: DEVNAME=/dev/kmsg
E: DEVPATH=/devices/virtual/mem/kmsg
E: MAJOR=1
E: MINOR=11
E: SUBSYSTEM=mem
```

2. at this point we can tag this device as being assigned to the snap, using the udev backend
3. with the output from udevadm info we have enough information to grant precise access to /run/udev. Eg, for kmsg above:
  /run/udev/data/c1:11 r,

Opengl is similar:
$ udevadm info /dev/dri/card0
P: /devices/pci0000:00/0000:00:02.0/drm/card0
N: dri/card0
E: DEVNAME=/dev/dri/card0
E: DEVPATH=/devices/pci0000:00/0000:00:02.0/drm/card0
E: DEVTYPE=drm_minor
E: ID_FOR_SEAT=drm-pci-0000_00_02_0
E: ID_PATH=pci-0000:00:02.0
E: ID_PATH_TAG=pci-0000_00_02_0
E: MAJOR=226
E: MINOR=0
E: SUBSYSTEM=drm
E: TAGS=:master-of-seat:seat:uaccess:
E: USEC_INITIALIZED=3062574

Add:
  /run/udev/data/c226:0 r,
  /sys/devices/pci0000:00/0000:00:02.0/{,**} r, # based on DEVPATH
  /run/udev/data/+pci:0000:00:02.0 r, # based on ID_PATH

We'll have to have some light per-interface mappings (eg, when ID_PATH is `pci-0000:00:02.0`, use `+pci:0000:00:02.0` in /run/data, but there might already be libraries and udev queries that would do this for us). We might also have to recursively descend directories in /sys/devices to realpath() the files so we know what to add, but that isn't hard to do either. All the pieces are there.